src/vcpkg/tools.cpp

```diff
@@ -62,82 +62,459 @@ namespace vcpkg
+                "54c8915c17610ff27b0cf2c1cf",
+                "node-v16.15.1-win-x64.7z",
+            },
+            {
+                "node",
+                "linux",
+                nullopt,
+                "16.15.1",
+                "node-v16.15.1-linux-x64/bin/node",
+                "https://nodejs.org/dist/v16.15.1/node-v16.15.1-linux-x64.tar.gz",
+                "5ad3b4b9caeaa8d31503efa99f5a593118a267dec9d4181d019732126ba248ce9a901207115b3f6b899eb5b3f0373c7f77ea95"
+                "cc92ac625cddf437ee9b8b8919",
+                "node-v16.15.1-linux-x64.tar.gz",
+            },
+            {
+                "node",
+                "osx",
+                nullopt,
+                "16.15.1",
+                "node-v16.15.1-darwin-x64/bin/node",
+                "https://nodejs.org/dist/v16.15.1/node-v16.15.1-darwin-x64.tar.gz",
+                "90d0612bbe5467b6cf385c91a68b8daad0057e3e0ccacea44567f5b95b14f7481cb79784185ab1463b4bd990e092ff0f910957"
+                "6d1a1934b84e1c816582929611",
+                "node-v16.15.1-darwin-x64.tar.gz",
+            }};
+
+        size_t default_tool = 0;
+        bool existsDefaultTool = false;
+        for (std::size_t i = 0; i < NUM_TOOLS; i++)
+        {
+            // if possible, find a tool with matching tool name, architecture, and os
+            // otherwise, choose a tool with only matching name and os and nullopt
+            // as architecture (this is the default tool)
+            const ArchToolData& d = tool_data[i];
+            if (d.tool == toolname && d.arch.has_value() && d.arch.value_or_exit(VCPKG_LINE_INFO) == arch && d.os == os)
+            {
+                return d;
+            }
+            else if (d.tool == toolname && d.os == os && !d.arch.has_value())
+            {
+                default_tool = i;
+                existsDefaultTool = true;
+            }
+        }
+        if (existsDefaultTool)
+        {
+            return tool_data[default_tool];
+        }
+
+        return nullopt;
+    }
+
+    static Optional<ToolData> get_tool_data(StringView tool)
+    {
+        auto hp = get_host_processor();
 #if defined(_WIN32)
-        return parse_tool_data_from_xml(XML, XML_PATH, tool, "windows");
+        auto data = get_raw_tool_data(tool, hp, "windows");
 #elif defined(__APPLE__)
-        return parse_tool_data_from_xml(XML, XML_PATH, tool, "osx");
+        auto data = get_raw_tool_data(tool, hp, "osx");
 #elif defined(__linux__)
-        return parse_tool_data_from_xml(XML, XML_PATH, tool, "linux");
+        auto data = get_raw_tool_data(tool, hp, "linux");
 #elif defined(__FreeBSD__)
-        return parse_tool_data_from_xml(XML, XML_PATH, tool, "freebsd");
+        auto data = get_raw_tool_data(tool, hp, "freebsd");
 #elif defined(__OpenBSD__)
-        return parse_tool_data_from_xml(XML, XML_PATH, tool, "openbsd");
+        auto data = get_raw_tool_data(tool, hp, "openbsd");
 #else
         return nullopt;
 #endif
```

Comment:
Since the OS constant is compiled into vcpkg, should we just not emit entries in the table which don't match the OS we expect and remove that part of the test? (It's OK if you don't want to do that, e.g. to make it easier to replace with json in the future)

Reply:
I think we should keep the table as it is. We should be moving everything to json in the near future.